pkg/object/autocertmanager/autocertmanager.go

```diff
@@ -0,0 +1,417 @@
+		return fmt.Errorf("at least one challenge type must be enabled")
+	}
+
+	for i := range spec.Domains {
+		d := &spec.Domains[i]
+
+		// convert to puny code to support Chinese or other unicode domain names
+		_, err := idna.Lookup.ToASCII(d.Name)
+		if err != nil && d.Name[0] == '*' {
+			_, err = idna.Lookup.ToASCII(d.Name[1:])
+		}
+		if err != nil {
+			return fmt.Errorf("domain name contains invalid characters: %s", d.Name)
+		}
+
+		if d.Name[0] != '*' {
+			continue
+		}
+
+		if !spec.EnableDNS01 {
+			return fmt.Errorf("find wildcard domain name but DNS-01 challenge is disabled: %s", d.Name)
+		}
+
+		if _, err := newDNSProvider(d); err != nil {
+			return fmt.Errorf("DNS provider configuration is invalid: %v", err)
+		}
+	}
+	return nil
+}
+
+// Zone returns the zone the domain belongs to
+func (spec *DomainSpec) Zone() string {
+	if spec.DNSProvider != nil {
+		return spec.DNSProvider["zone"]
+	}
+	return ""
+}
+
+func init() {
+	supervisor.Register(&AutoCertManager{})
+}
+
+// Category returns the category of AutoCertManager.
+func (acm *AutoCertManager) Category() supervisor.ObjectCategory {
+	return Category
+}
+
+// Kind return the kind of AutoCertManager.
+func (acm *AutoCertManager) Kind() string {
+	return Kind
+}
+
+// DefaultSpec returns the default spec of AutoCertManager.
+func (acm *AutoCertManager) DefaultSpec() interface{} {
+	return &Spec{
+		DirectoryURL:    acme.LetsEncryptURL,
+		RenewBefore:     "720h",
+		EnableHTTP01:    true,
+		EnableTLSALPN01: true,
+		EnableDNS01:     true,
+	}
+}
+
+// Init initializes AutoCertManager.
+func (acm *AutoCertManager) Init(superSpec *supervisor.Spec) {
+	acm.superSpec = superSpec
+	acm.spec = superSpec.ObjectSpec().(*Spec)
+	acm.super = superSpec.Super()
+
+	// TODO: remove this check after converting AutoCertManager to a system controller
+	if globalACM.Load() != nil {
+		logger.Warnf("an AutoCertManager instance is already exist")
+	}
+
+	acm.reload()
+}
+
+// Inherit inherits previous generation of AutoCertManager.
+func (acm *AutoCertManager) Inherit(superSpec *supervisor.Spec, previousGeneration supervisor.Object) {
+	acm.superSpec = superSpec
+	acm.spec = superSpec.ObjectSpec().(*Spec)
+	acm.super = superSpec.Super()
+
+	acm.reload()
+	previousGeneration.(*AutoCertManager).Close()
+}
+
+func (acm *AutoCertManager) findDomain(name string, exactMatch bool) *Domain {
+	// compare one by one, if many domains are configured, we need to refactor
+	// this function to improve performance.
+	for i := range acm.domains {
+		domain := &acm.domains[i]
+		if name == domain.Name {
+			return domain
+		}
+		if exactMatch || !domain.isWildcard() {
+			continue
+		}
+		if strings.HasSuffix(name, domain.Name[1:]) {
+			return domain
+		}
+	}
+	return nil
+}
+
+func (acm *AutoCertManager) reload() {
+	acm.stopCtx, acm.cancel = context.WithCancel(context.Background())
+	acm.storage = newStorage(acm.super.Cluster())
+
+	acm.renewBefore, _ = time.ParseDuration(acm.spec.RenewBefore)
```

Comment:
log an error here? maybe custom want their renew time but write it in wrong format...

Reply:
Cool! The code is a little bit complicated, I miss this part....